transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu

```diff
@@ -944,4 +957,178 @@ void dispatch_scaled_masked_softmax_backward(
     }
 }
 
-#endif  // TRANSFORMER_ENGINE_SCALED_MASKED_SOFTMAX_H_
+
+void scaled_softmax_forward_cuda(
+    const Tensor &input,
+    Tensor *softmax_results,
+    float scale_factor,
+    cudaStream_t stream) {
+
+    const int batches = input.shape[0];
+    const int attn_heads = input.shape[1];
+    const int query_seq_len = input.shape[2];
+    const int key_seq_len = input.shape[3];
+
+    TRANSFORMER_ENGINE_TYPE_SWITCH_16BIT(input.dtype, softmax_type,
```

Comment:
Does this softmax only work for 16-bit values? That is some pretty serious limitation :-(.

Reply:
Yes, currently this restriction is present. For `FP32`, the framework softmax is used in TE `softmax.py`.